src/webhook_launcher/app/models.py

```diff
@@ -392,24 +399,127 @@ def clean(self, exclude=None):
                     (project, self.user)
                 )
 
-    def trigger_build(self):
-        if not self.lsr:
-            mylsr, created = LastSeenRevision.objects.get_or_create(
-                mapping=self,
+    def trigger_build(self, user=None, tag=None, force=False):
+        if not self.pk:
+            raise RuntimeError(
+                "trigger_build() on unsaved WebHookMapping"
+            )
+
+        # Only fire for projects which allow webhooks. We can't just
+        # rely on validation since a Project may forbid hooks after
+        # the hook was created
+        if self.project_disabled:
+            print "Project has build disabled"
+            return
+
+        handled = self.lsr.handled and self.lsr.tag == tag and not force
+        if handled:
+            print "build already handled, skipping"
+        build = self.build and self.mapped and not handled
+        qp = None
+        if user is None:
+            user = self.user.username
+
+        if build:
+            if tag:
+                self.lsr.tag = tag
```

Comment:
I'm not sure that we should be overwriting the 'tag' in an LSR?
Should that not be immutable once it's come from the payload?